fem/intrules_cut.cpp

```diff
@@ -31,6 +31,174 @@ void CutIntegrationRules::SetLevelSetProjectionOrder(int order)
+         ip.Set2w(q.nodes[i].x(0),q.nodes[i].x(1),q.nodes[i].w);
+      }
+   }
+   else
+   {
+      LevelSet3D ls(pe,lsvec);
+      auto q = Algoim::quadGen<3>(ls,Algoim::BoundingBox<real_t,3>(0.0,1.0),
+                                  -1, -1, np1d);
+
+      result.SetSize(q.nodes.size());
+      result.SetOrder(CutIntegrationRules::Order);
+      for (size_t i=0; i<q.nodes.size(); i++)
+      {
+         IntegrationPoint& ip=result.IntPoint(i);
+         ip.Set(q.nodes[i].x(0),q.nodes[i].x(1),q.nodes[i].x(2),q.nodes[i].w);
+      }
+   }
+
+}
+
+void AlgoimIntegrationRules::GetSurfaceWeights(ElementTransformation &Tr,
+                                               const IntegrationRule &sir,
+                                               Vector &weights)
+{
+   GenerateLSVector(Tr,LvlSet);
+
+   DenseMatrix bmat; // gradients of the shape functions in isoparametric space
+   DenseMatrix pmat; // gradients of the shape functions in physical space
+   Vector inormal; // normal to the level set in isoparametric space
+   Vector tnormal; // normal to the level set in physical space
+   bmat.SetSize(pe->GetDof(),pe->GetDim());
+   pmat.SetSize(pe->GetDof(),pe->GetDim());
+   inormal.SetSize(pe->GetDim());
+   tnormal.SetSize(pe->GetDim());
+
+   weights.SetSize(sir.GetNPoints());
+
+   for (int j = 0; j < sir.GetNPoints(); j++)
+   {
+      const IntegrationPoint &ip = sir.IntPoint(j);
+      Tr.SetIntPoint(&ip);
+      pe->CalcDShape(ip,bmat);
+      Mult(bmat, Tr.InverseJacobian(), pmat);
+      // compute the normal to the LS in isoparametric space
+      bmat.MultTranspose(lsvec,inormal);
+      // compute the normal to the LS in physical space
+      pmat.MultTranspose(lsvec,tnormal);
+      weights[j]= tnormal.Norml2() / inormal.Norml2();
+   }
+
+}
+
+void AlgoimIntegrationRules::GenerateLSVector(ElementTransformation &Tr,
+                                              Coefficient* lvlset)
+{
+   //check if the coefficient is already projected
+   if (currentElementNo==Tr.ElementNo)
+   {
+      if (currentLvlSet==lvlset)
+      {
+         if (currentGeometry==Tr.GetGeometryType())
+         {
+            return;
+         }
+      }
+   }
+
+   currentElementNo=Tr.ElementNo;
+
+   if (currentGeometry!=Tr.GetGeometryType())
+   {
+      delete le;
+      delete pe;
+      currentGeometry=Tr.GetGeometryType();
+      if (Tr.GetGeometryType()==Geometry::Type::SQUARE)
+      {
+         pe=new H1Pos_QuadrilateralElement(lsOrder);
+         le=new H1_QuadrilateralElement(lsOrder);
+      }
+      else if (Tr.GetGeometryType()==Geometry::Type::CUBE)
+      {
+         pe=new H1Pos_HexahedronElement(lsOrder);
+         le=new H1_HexahedronElement(lsOrder);
+
+      }
+      else
+      {
+         MFEM_ABORT("Currently MFEM + Algoim supports only quads and hexes.");
+      }
+
+      T.SetSize(pe->GetDof());
+      pe->Project(*le,Tr,T);
+      //The transformation matrix depends only on the geometry
+      // for change of basis
```

Comment:
```suggestion
      //The transformation matrix depends only on geometry for change of basis.
```
Could format into one line.

Reply:
fixed